Better Stats API

Machine-Maker opened the pull request:
Replaces https://github.com/PaperMC/Paper/pull/5691

Machine-Maker committed:
Better Stats API

Warriorrrr commented on paper-api/src/main/java/io/papermc/paper/statistic/Statistic.java:
```diff
@@ -0,0 +1,27 @@
+ * @param <S> stat (one of {@link org.bukkit.entity.EntityType}, {@link org.bukkit.Material} or {@link CustomStatistic}).
+ */
+@ApiStatus.NonExtendable
+public interface Statistic<S> extends Criteria {
```
Perhaps add an example here in the javadoc about how to obtain an instance of this via the StatisticType or CustomStatistic classes, since this is the first place that people who see the deprecations will be looking at.

Warriorrrr commented on paper-api/src/main/java/org/bukkit/event/player/PlayerStatisticIncrementEvent.java:
```diff
@@ -108,8 +98,10 @@ public EntityType getEntityType() {
+     * @deprecated use {@link #getStat()}
      */
     @Nullable
+    @Deprecated(since = "1.21,4")
```
contains a comma instead of a period

lynxplay committed:
1.21.6 dev

Co-authored-by: Bjarne Koll <git@lynxplay.dev>
Co-authored-by: Jake Potrebic <jake.m.potrebic@gmail.com>
Co-authored-by: Jason Penilla <11360596+jpenilla@users.noreply.github.com>
Co-authored-by: Lulu13022002 <41980282+Lulu13022002@users.noreply.github.com>
Co-authored-by: Noah van der Aa <ndvdaa@gmail.com>
Co-authored-by: Owen1212055 <23108066+Owen1212055@users.noreply.github.com>
Co-authored-by: Spottedleaf <Spottedleaf@users.noreply.github.com>

Lulu13022002 committed:
tweak mob goal keys, generate missing api

Rename the mob goal keys to avoid future conflict and a more
consistent scheme <mob>_<name> except for generic enough types.
And fix an edge case with the tempt mob goal key not selecting the
right constructor.

Lulu13022002 committed:
Revert "fix sponge-absorb deleting chest content (#12647)"

This reverts commit 320f25cb04f183c18c16595100897622c88ce073.

Lulu13022002 committed:
generate

Lulu13022002 committed:
support parameterized type for generated tag keys

jpenilla committed:
Allow cache writes on main and dev branches

JRoy committed:
Throw exception for CraftMagicNumbers#getMappingsVersion (#12676)

jpenilla committed:
[ci skip] Fix test results event file
